test --doc can't find dependency of dependency with a proc macro when passing --target

I was trying to add CI to a proc macro crate of mine, but the CI passes `--target` explicitly even though it's unneeded. To my surprise the test failed https://travis-ci.org/parched/runtime-target-feature-rs/jobs/246862218

I've tried to create a minimal repro below, possibly there are some extra things that aren't needed.

Cargo.toml
```
...

[lib]
proc-macro = true

[dependencies]
syn = "0.11"
```

`src/lib.rs`
```
#![feature(proc_macro)]

extern crate proc_macro;
extern crate syn;

use proc_macro::TokenStream;

#[proc_macro_attribute]
pub fn test(_: TokenStream, function: TokenStream) -> TokenStream {
    function
}
```


```
$ cargo test --doc --target x86_64-unknown-linux-gnu
   Compiling quote v0.3.15
   Compiling unicode-xid v0.0.4
   Compiling synom v0.11.3
   Compiling syn v0.11.11
   Compiling runtime-target-feature v0.1.0 (file:///home/james/junk/cargo-dep-test)
    Finished dev [unoptimized + debuginfo] target(s) in 5.35 secs
   Doc-tests runtime-target-feature
error[E0463]: can't find crate for `quote` which `syn` depends on
 --> /home/james/junk/cargo-dep-test/src/lib.rs:4:1
  |
4 | extern crate syn;
  | ^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error(s)

thread '<unnamed>' panicked at 'Box<Any>', src/librustc_errors/lib.rs:512
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: test failed, to rerun pass '--doc'
```

without explicitly passing my native target
```
$ cargo test --doc
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
   Doc-tests runtime-target-feature

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

```
$ rustc --version
rustc 1.20.0-dev (ec5c2e8fb 2017-06-20)
$ cargo --version
cargo 0.20.0-nightly (0d0f3baad 2017-06-13)
```

--- a/src/cargo/ops/cargo_rustc/compilation.rs
+++ b/src/cargo/ops/cargo_rustc/compilation.rs
@@ -35,10 +35,6 @@ pub struct Compilation<'cfg> {
     /// Output directory for the rust host dependencies.
     pub host_deps_output: PathBuf,
 
-    /// Library search path for compiler plugins and build scripts
-    /// which have dynamic dependencies.
-    pub plugins_dylib_path: PathBuf,
-
     /// The path to rustc's own libstd
     pub host_dylib_path: Option<PathBuf>,
 
@@ -69,7 +65,6 @@ impl<'cfg> Compilation<'cfg> {
             root_output: PathBuf::from("/"),
             deps_output: PathBuf::from("/"),
             host_deps_output: PathBuf::from("/"),
-            plugins_dylib_path: PathBuf::from("/"),
             host_dylib_path: None,
             target_dylib_path: None,
             tests: Vec::new(),
@@ -129,7 +124,7 @@ impl<'cfg> Compilation<'cfg> {
                 -> CargoResult<ProcessBuilder> {
 
         let mut search_path = if is_host {
-            let mut search_path = vec![self.plugins_dylib_path.clone()];
+            let mut search_path = vec![self.host_deps_output.clone()];
             search_path.extend(self.host_dylib_path.clone());
             search_path
         } else {

--- a/src/cargo/ops/cargo_rustc/compilation.rs
+++ b/src/cargo/ops/cargo_rustc/compilation.rs
@@ -28,8 +28,12 @@ pub struct Compilation<'cfg> {
     /// Root output directory (for the local package's artifacts)
     pub root_output: PathBuf,
 
-    /// Output directory for rust dependencies
+    /// Output directory for rust dependencies.
+    /// May be for the host or for a specific target.
     pub deps_output: PathBuf,
+
+    /// Output directory for the rust host dependencies.
+    pub host_deps_output: PathBuf,
 
     /// Library search path for compiler plugins and build scripts
     /// which have dynamic dependencies.
@@ -64,6 +68,7 @@ impl<'cfg> Compilation<'cfg> {
             native_dirs: HashSet::new(),  // TODO: deprecated, remove
             root_output: PathBuf::from("/"),
             deps_output: PathBuf::from("/"),
+            host_deps_output: PathBuf::from("/"),
             plugins_dylib_path: PathBuf::from("/"),
             host_dylib_path: None,
             target_dylib_path: None,

--- a/src/cargo/ops/cargo_rustc/context.rs
+++ b/src/cargo/ops/cargo_rustc/context.rs
@@ -176,6 +176,7 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         }
 
         self.compilation.plugins_dylib_path = self.host.deps().to_path_buf();
+        self.compilation.host_deps_output = self.host.deps().to_path_buf();
 
         let layout = self.target.as_ref().unwrap_or(&self.host);
         self.compilation.root_output = layout.dest().to_path_buf();

--- a/src/cargo/ops/cargo_rustc/context.rs
+++ b/src/cargo/ops/cargo_rustc/context.rs
@@ -175,7 +175,6 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
             None => {}
         }
 
-        self.compilation.plugins_dylib_path = self.host.deps().to_path_buf();
         self.compilation.host_deps_output = self.host.deps().to_path_buf();
 
         let layout = self.target.as_ref().unwrap_or(&self.host);

--- a/src/cargo/ops/cargo_test.rs
+++ b/src/cargo/ops/cargo_test.rs
@@ -126,7 +126,6 @@ fn run_doc_tests(options: &TestOptions,
                  -> CargoResult<(Test, Vec<ProcessError>)> {
     let mut errors = Vec::new();
     let config = options.compile_opts.config;
-    let host_deps = find_host_deps(options, compilation);
 
     // We don't build/rust doctests if target != host
     if config.rustc()?.host != compilation.target {
@@ -145,15 +144,20 @@ fn run_doc_tests(options: &TestOptions,
             p.arg("--test").arg(lib)
              .arg("--crate-name").arg(&crate_name);
 
-            p.arg("-L").arg(&host_deps);
-
             for &rust_dep in &[&compilation.deps_output] {
                 let mut arg = OsString::from("dependency=");
                 arg.push(rust_dep);
                 p.arg("-L").arg(arg);
             }
+
             for native_dep in compilation.native_dirs.iter() {
                 p.arg("-L").arg(native_dep);
+            }
+
+            for &host_rust_dep in &[&compilation.host_deps_output] {
+                let mut arg = OsString::from("dependency=");
+                arg.push(host_rust_dep);
+                p.arg("-L").arg(arg);
             }
 
             for arg in test_args {
@@ -200,25 +204,4 @@ fn run_doc_tests(options: &TestOptions,
         }
     }
     Ok((Test::Doc, errors))
-}
-
-fn find_host_deps(options: &TestOptions, compilation: &Compilation) -> OsString {
-    let build_type = if options.compile_opts.release { "release" } else { "debug" };
-    let mut dir = compilation.root_output.clone();
-
-    // first pop off the build_type
-    dir.pop();
-    // if we see the target next, pop it off
-    let target: &OsStr = compilation.target.as_ref();
-    if dir.file_name().unwrap() == target {
-        dir.pop();
-    }
-    // push the build_type back on
-    dir.push(build_type);
-    // and we are looking for the deps directory
-    dir.push("deps");
-
-    let mut host_deps = OsString::from("dependency=");
-    host_deps.push(dir);
-    host_deps
 }

--- a/src/cargo/ops/cargo_test.rs
+++ b/src/cargo/ops/cargo_test.rs
@@ -126,6 +126,7 @@ fn run_doc_tests(options: &TestOptions,
                  -> CargoResult<(Test, Vec<ProcessError>)> {
     let mut errors = Vec::new();
     let config = options.compile_opts.config;
+    let host_deps = find_host_deps(options, compilation);
 
     // We don't build/rust doctests if target != host
     if config.rustc()?.host != compilation.target {
@@ -143,6 +144,8 @@ fn run_doc_tests(options: &TestOptions,
             let mut p = compilation.rustdoc_process(package)?;
             p.arg("--test").arg(lib)
              .arg("--crate-name").arg(&crate_name);
+
+            p.arg("-L").arg(&host_deps);
 
             for &rust_dep in &[&compilation.deps_output] {
                 let mut arg = OsString::from("dependency=");
@@ -197,4 +200,25 @@ fn run_doc_tests(options: &TestOptions,
         }
     }
     Ok((Test::Doc, errors))
+}
+
+fn find_host_deps(options: &TestOptions, compilation: &Compilation) -> OsString {
+    let build_type = if options.compile_opts.release { "release" } else { "debug" };
+    let mut dir = compilation.root_output.clone();
+
+    // first pop off the build_type
+    dir.pop();
+    // if we see the target next, pop it off
+    let target: &OsStr = compilation.target.as_ref();
+    if dir.file_name().unwrap() == target {
+        dir.pop();
+    }
+    // push the build_type back on
+    dir.push(build_type);
+    // and we are looking for the deps directory
+    dir.push("deps");
+
+    let mut host_deps = OsString::from("dependency=");
+    host_deps.push(dir);
+    host_deps
 }

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -2816,10 +2816,10 @@ fn find_dependency_of_proc_macro_dependency_with_target() {
             proc-macro = true
 
             [dependencies]
-            base64 = "^0.6"
+            dep_of_proc_macro_dep = "^0.1"
         "#)
         .file("proc_macro_dep/src/lib.rs", r#"
-            extern crate base64;
+            extern crate dep_of_proc_macro_dep;
             extern crate proc_macro;
             use proc_macro::TokenStream;
 
@@ -2828,6 +2828,7 @@ fn find_dependency_of_proc_macro_dependency_with_target() {
                 "".parse().unwrap()
             }
         "#);
+        Package::new("dep_of_proc_macro_dep", "0.1.0").publish();
         workspace.build();
     assert_that(workspace.cargo("test").arg("--all").arg("--target").arg(rustc_host()),
                 execs().with_status(0));

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -2816,10 +2816,10 @@ fn find_dependency_of_proc_macro_dependency_with_target() {
             proc-macro = true
 
             [dependencies]
-            dep_of_proc_macro_dep = "^0.1"
+            bar = "^0.1"
         "#)
         .file("proc_macro_dep/src/lib.rs", r#"
-            extern crate dep_of_proc_macro_dep;
+            extern crate bar;
             extern crate proc_macro;
             use proc_macro::TokenStream;
 
@@ -2828,7 +2828,11 @@ fn find_dependency_of_proc_macro_dependency_with_target() {
                 "".parse().unwrap()
             }
         "#);
-    Package::new("dep_of_proc_macro_dep", "0.1.0").publish();
+    Package::new("foo", "0.1.0").publish();
+    Package::new("bar", "0.1.0")
+        .dep("foo", "0.1")
+        .file("src/lib.rs", "extern crate foo;")
+        .publish();
     workspace.build();
     assert_that(workspace.cargo("test").arg("--all").arg("--target").arg(rustc_host()),
                 execs().with_status(0));

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -6,7 +6,7 @@ use std::fs::File;
 use std::io::prelude::*;
 use std::str;
 
-use cargotest::{sleep_ms, is_nightly};
+use cargotest::{sleep_ms, is_nightly, rustc_host};
 use cargotest::support::{project, execs, basic_bin_manifest, basic_lib_manifest, cargo_exe};
 use cargotest::support::paths::CargoPathExt;
 use cargotest::support::registry::Package;
@@ -2780,5 +2780,55 @@ fn publish_a_crate_without_tests() {
 
     assert_that(p.cargo("test"), execs().with_status(0));
     assert_that(p.cargo("test").arg("--package").arg("testless"),
+                execs().with_status(0));
+}
+
+#[test]
+fn find_dependency_of_proc_macro_dependency_with_target() {
+    let workspace = project("workspace")
+        .file("Cargo.toml", r#"
+            [workspace]
+            members = ["root", "proc_macro_dep"]
+        "#)
+        .file("root/Cargo.toml", r#"
+            [project]
+            name = "root"
+            version = "0.1.0"
+            authors = []
+
+            [dependencies]
+            proc_macro_dep = { path = "../proc_macro_dep" }
+        "#)
+        .file("root/src/lib.rs", r#"
+            #[macro_use]
+            extern crate proc_macro_dep;
+
+            #[derive(Noop)]
+            pub struct X;
+        "#)
+        .file("proc_macro_dep/Cargo.toml", r#"
+            [project]
+            name = "proc_macro_dep"
+            version = "0.1.0"
+            authors = []
+
+            [lib]
+            proc-macro = true
+
+            [dependencies]
+            base64 = "^0.6"
+        "#)
+        .file("proc_macro_dep/src/lib.rs", r#"
+            extern crate base64;
+            extern crate proc_macro;
+            use proc_macro::TokenStream;
+
+            #[proc_macro_derive(Noop)]
+            pub fn noop(_input: TokenStream) -> TokenStream {
+                "".parse().unwrap()
+            }
+        "#);
+        workspace.build();
+    assert_that(workspace.cargo("test").arg("--all").arg("--target").arg(rustc_host()),
                 execs().with_status(0));
 }

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -2828,8 +2828,8 @@ fn find_dependency_of_proc_macro_dependency_with_target() {
                 "".parse().unwrap()
             }
         "#);
-        Package::new("dep_of_proc_macro_dep", "0.1.0").publish();
-        workspace.build();
+    Package::new("dep_of_proc_macro_dep", "0.1.0").publish();
+    workspace.build();
     assert_that(workspace.cargo("test").arg("--all").arg("--target").arg(rustc_host()),
                 execs().with_status(0));
 }

